### PR TITLE
[IGA] Load condition

### PR DESCRIPTION
**Pull request by tteschemacher**
Hi,

this is the condition that I am using for application of loads with IGA. However, could be used also in standard FEM or MPM. It's kept simple and the issues are resolved within the geometry.

The test are combined with the geometry and I will add them later when everything is merged.

### CODE REVIEW

File: applications/IgaApplication/custom_conditions/load_condition.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+                const double d_weight = integration_weight * determinat_jacobian_vector[point_number];
+
+                // Split only due to different existing variable names
+                // No check included, which checks correctness of variable
+
+                // Point loads
+                if (this->Has(POINT_LOAD))
+                {
+                    const array_1d<double, 3> point_load = this->GetValue(POINT_LOAD);
```
Comment (loumalouomega) [inline applications/IgaApplication/custom_conditions/load_condition.cpp]:
```suggestion
                    const array_1d<double, 3>& point_load = this->GetValue(POINT_LOAD);
```
And therefore should be a r_point_load

File: applications/IgaApplication/custom_conditions/load_condition.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                        f[index]     += point_load[0] * r_N(point_number, i);
+                        f[index + 1] += point_load[1] * r_N(point_number, i);
+                        f[index + 2] += point_load[2] * r_N(point_number, i);
+                    }
+                }
+
+                // Line loads
+                if (this->Has(LINE_LOAD))
+                {
+                    const array_1d<double, 3> line_load = this->GetValue(LINE_LOAD);
```
Comment (loumalouomega) [inline applications/IgaApplication/custom_conditions/load_condition.cpp]:
```suggestion
                    const array_1d<double, 3>& line_load = this->GetValue(LINE_LOAD);
```
Remember to put a reference in anything more complex than a double

File: applications/IgaApplication/custom_conditions/load_condition.cpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                        f[index]     += line_load[0] * r_N(point_number, i) * d_weight;
+                        f[index + 1] += line_load[1] * r_N(point_number, i) * d_weight;
+                        f[index + 2] += line_load[2] * r_N(point_number, i) * d_weight;
+                    }
+                }
+
+                // Surface loads
+                if (this->Has(SURFACE_LOAD))
+                {
+                    const array_1d<double, 3> surface_load = this->GetValue(SURFACE_LOAD);
```
Comment (loumalouomega) [inline applications/IgaApplication/custom_conditions/load_condition.cpp]:
```suggestion
                    const array_1d<double, 3>& surface_load = this->GetValue(SURFACE_LOAD);
```

File: applications/IgaApplication/custom_conditions/load_condition.cpp [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+                        f[index]     += surface_load[0] * r_N(point_number, i) * d_weight;
+                        f[index + 1] += surface_load[1] * r_N(point_number, i) * d_weight;
+                        f[index + 2] += surface_load[2] * r_N(point_number, i) * d_weight;
+                    }
+                }
+
+                // Pressure loads
+                if (this->Has(PRESSURE))
+                {
+                    double pressure = this->GetValue(PRESSURE);
```
Comment (loumalouomega) [inline applications/IgaApplication/custom_conditions/load_condition.cpp]:
```suggestion
                    const double pressure = this->GetValue(PRESSURE);
```

**Review by loumalouomega — COMMENTED** [review]
Some comments. I don't understand which is the problem with the structural base load condition?

**Comment by loumalouomega** [comment]
You are trying to merge in another branch, not int the master

**Comment by tteschemacher** [comment]
Thanks Vicente!

Well, the major difference is that, here we don't read the variables from nodes/dofs, here we have to apply the values with the process add_variable_to_condition. Thus, it is going the other way. This is advantageous, especially for embedded structures.

**Comment by loumalouomega** [comment]
> Thanks Vicente!
> 
> Well, the major difference is that, here we don't read the variables from nodes/dofs, here we have to apply the values with the process add_variable_to_condition. Thus, it is going the other way. This is advantageous, especially for embedded structures.

Our implementation also reads from the condition database (additionally to the nodes, which i don't support for the humongous increase in memory consumption)

https://github.com/KratosMultiphysics/Kratos/blob/702b361dd77004b2d3c3075d6c42070df1754727/applications/StructuralMechanicsApplication/custom_conditions/surface_load_condition_3d.cpp#L260

**Comment by tteschemacher** [comment]
Yes, that is true. I need the norm to be in master first.

**Comment by tteschemacher** [comment]
Actually, you're right. This condition should work for our cases, too.

**Comment by tteschemacher** [comment]
Only difference is, that this one is more generic, independent on the used geometry. Can be surface, embedded line,... That is actually why I needed the normal in the core.

**Comment by loumalouomega** [comment]
> Actually, you're right. This condition should work for our cases, too.

imagen

**Comment by loumalouomega** [comment]
> Only difference is, that this one is more generic, independent on the used geometry. Can be surface, embedded line,... That is actually why I needed the normal in the core.

Maybe is clever to move this code to base_load_condition and keep the current implementation for derived ones (which have less overhead)

**Comment by tteschemacher** [comment]
That would make a lot of sence to me!

**Comment by philbucher** [comment]
what is the status here?

**Comment by tteschemacher** [comment]
I possible I would like to merge that. As soon as the one in Structural Mechanics is done we can think about removing it.

**Review by philbucher — APPROVED** [review]
mergen Sie

**Comment by philbucher** [comment]
seems like you have to merge master

**Comment by tteschemacher** [comment]
Grazie!

**Pull request merged**